connection.cpp

```diff
@@ -87,28 +86,72 @@ Connection::~Connection()
     delete d;
 }
 
+void Connection::setServer(const QString &domain, int tlsPort)
+{
+    d->data->setHost(domain);
+    d->data->setPort(tlsPort);
+}
+
 void Connection::resolveServer(const QString& domain)
 {
+    qDebug() << "ResolveServer:" << domain;
+
     // Find the Matrix server for the given domain.
-    QScopedPointer<QDnsLookup, QScopedPointerDeleteLater> dns { new QDnsLookup() };
+    dns = new QDnsLookup();
+
+    connect(dns, SIGNAL(finished()),this, SLOT(handleServerSRV()));
+
     dns->setType(QDnsLookup::SRV);
     dns->setName("_matrix._tcp." + domain);
-
     dns->lookup();
-    connect(dns.data(), &QDnsLookup::finished, [&]() {
-        // Check the lookup succeeded.
-        if (dns->error() != QDnsLookup::NoError ||
-                dns->serviceRecords().isEmpty()) {
-            emit resolveError("DNS lookup failed");
-            return;
-        }
+}
 
-        // Handle the results.
-        auto record = dns->serviceRecords().front();
-        d->data->setHost(record.target());
-        d->data->setPort(record.port());
-        emit resolved();
-    });
+void Connection::handleServerSRV()
+{
+    qDebug() << "dnslookup finished";
+
+    disconnect(dns, SIGNAL(finished()),this, SLOT(handleServerSRV()));
+
+    // Check the lookup succeeded.
+    if (dns->error() != QDnsLookup::NoError ) {
+        connect(dns, SIGNAL(finished()),this, SLOT(handleServerA()));
+
+        dns->setType(QDnsLookup::A);
+        dns->lookup();
+        qDebug() << "Resolve SRV Failed";
+        return;
+    }
+
+    qDebug() << "Resolved:" << dns->name();
+
+    // Handle the results.
+    auto record = dns->serviceRecords().front();
+
+    this->setServer(record.target(), record.port());
+    emit resolved();
+    dns->deleteLater();
+}
+
+void Connection::handleServerA()
```

Comment:
Don't you feel that you're duplicating code here? Maybe try to pass everything necessary as parameters? You can do it, e.g., by passing not the function pointer to `connect`, but a lambda that will call some more generic `handleDnsRecord()` with necessary parameters - including a recursive `connect` from itself. As for the parameters, I'd make an array or a list of structures that store information on what to try: a record to request and a port to set by default. That seems to be all that needs to be parameterised but I may be mistaking.